### ISSUE TITLE
Separate "generate" and "import" steps of install command

### ISSUE THREAD

**Issue by annulen** [conan-io/conan#632]
For Qt project it would be great to have ability to run `install` command without doing imports, and then do imports as a separate stage (with different output directories for running tests and for packaging)

**Comment by memsharded** [comment]
Agree, this could be made opt-in (in order not to break existing usage), with a command line argument as ``--no-import`` to ``conan install`` command.

Doing import in a separate stage is basically a full ``conan install`` command, as the imports need the full dependency graph, and existing binary packages. If they have been previously retrieved, that part of the ``conan install`` will be fast, if not, it will require them to be able to import anything. So I would say that the ``--no-import`` argument could solve the issue, please tell me. Thanks!

**Comment by annulen** [comment]
Full `conan install` generates `conanbuildinfo` files, and if we want to add 3rd party libraries to results of `make install` (with imports), we don't want to have `conanbuildinfo` files in the installation image

**Comment by memsharded** [comment]
Just an idea. You could have a custom ``conanfile.txt`` for this purpose, something like:

**Comment by annulen** [comment]
Indeed, this can work as a workaround, however it obscures things (as `imports` depend on to `requires` and their `options`)

**Comment by annulen** [comment]
I haven't looked at the implementation yet but is it really hard to run just `imports()`?

**Comment by memsharded** [comment]
The problem with ``imports()`` is that it doesn't use the full local cache. It needs to know all the dependencies (the dependency graph) of the project, evaluate those dependencies with the current settings and options to obtain the matching binaries IDs (you don't want to copy artifacts for other different settings), the matching binaries have to exist in the local cache, and finally, copy the artifacts from the local cache to the user folder. 

That is basically a ``conan install`` command. So it is very easy to just skip the last ``imports`` functionality, but running ``imports()`` without having the dependency tree would require a different approach. 

Maybe we could use a generated ``conanbuildinfo.txt`` file, generated with the ``txt`` generator at ``conan install`` time, in a similar way that it is used for local ``source, build and package`` commands running in the user space. Another approach would be an argument to skip local generation of files. We can have a look, but indeed would be more complicated to implement, so I will remove the "easy" label while we check that. Thanks for the feedback!

**Comment by annulen** [comment]
>So it is very easy to just skip the last imports functionality, but running imports() without having the dependency tree would require a different approach. 

Hmm, avoiding dependency tree doesn't sound right. I think we need full install without running generators, instead of running only imports.

**Comment by memsharded** [comment]
Hi @annulen,

I have been checking this, and implemented a few things. It seems possible to implement a ``conan imports`` command that uses a generated ``conanbuildinfo.txt``, because that file encodes all the information necessary for performing the ``imports``.

I am wondering about the flow. The ``imports()``  method of a conanfile is used to retrieve exes, shared libs and other data from its dependencies. So something like:

```bash
$ conan install Qt/X.Y@user/channel
$ conan imports Qt/X.Y@user/channel --dest="."
```
With the ``conanfile.py`` for such Qt package, having the typical implementatio of imports.

This will only be able to copy the possible dependencies of Qt to the current path, but not Qt itself.

The flow would be different, if what we have is a local ``conanfile.txt`` (or .py), that "requires" Qt/X.Y@user/channel, and defines some ``imports`` in it, then:

```bash
$ conan install .
$ conan imports .
```

Might be able to copy both the artifacts from Qt and its dependencies.

What would be your flow? How do you intend to use this feature? You might want to have a look at this feature request, which I think is closely related: https://github.com/conan-io/conan/issues/636

Thanks very much for your feedback!

**Comment by annulen** [comment]
Intention is to run "imports" to add 3rd party libraries to installation of Qt module as produces by "make install", to create installation image without unresolved dependencies, that can be given to end users.

make install cannot copy these libraries because it works under assumption that libraries are taken from some "system" paths and will stay there. So, there are 2 solutions: 1) let conan to import libraries right into installation prefix; 2) make additional scripts for CI system to "install" those libraries manually

    conan imports -f $conanfile --dest=$prefix

should be enough for this workflow

**Comment by memsharded** [comment]
This feature has been implemented by https://github.com/conan-io/conan/pull/686, will be released in next 0.16

**Comment by memsharded** [comment]
Released in 0.16
